### ISSUE TITLE
Disallow invalid subnet/Port-Address

### ISSUE THREAD

**Issue by Firionus** [Trangar/artnet_protocol#9]
It is currently possible to produce ArtDmx packets with subnets above the 15-bit limit of 32_767:
```Rust
let artnet_protocol_packet = ArtCommand::Output(Output {
        length: 512,
        data: vec![0xff; 512],
        subnet: 32_768, //not allowed
        ..Output::default()
    });
```
See the [Art-Net Spec](https://artisticlicence.com/WebSiteMaster/User%20Guides/art-net.pdf) on page 5 where the Port-Address is defined to be a 15-bit number or a u16 with the highest bit being 0. 

# How to fix?

I'm new to Rust and not sure how to approach this issue and #7 in the context of your existing code. Especially the macro-driven approach seems opaque to me. 

To get a better grip on how I would do it (and to do my first Rust project), I implemented [my own Art-Net packer](https://github.com/Firionus/artnet_packer). Notable differences include: 
- Custom types for Port Address and DMX universe data that handle validation upon creation
- A stateful conversion from DMX data to the Art-Net packets. This is how I was able to implement suppression (not sending the same packet too often). It could also produce sequence numbers in the future. These stateful features are part of the Art-Net specification and I think therefore these should be handled by an Art-Net crate. 
- some more unit tests

I don't want to implement and maintain my own Art-Net crate when you already have a crate that is used by people. But I'd love if you could take a look at my repo (there's tests and docs, only need to build with `cargo doc`) and see how we could integrate some of my ideas into your crate. I can help, but probably need some guidance.

**Comment by VictorKoenders** [comment]
> I don't want to implement and maintain my own Art-Net crate when you already have a crate that is used by people.

I'm not convinced many people use this crate. It's something I've put together a couple of years ago for a project and after that I've never used artnet again. If you want to create and publish your own crate, feel free!

How I'd fix this particular issue is create a struct and implement the [Convertable](https://github.com/Trangar/artnet_protocol/blob/master/src/convert.rs#L6-L11) trait on that.

You can do a conversion from u16 with methods
- use `subnet: Subnet::new(1234)` and validate in the `fn new` if the given value is valid
- use `impl From<u16> for Subnet`, which would make the developer call it with `subnet: 1234.into()`, make `Convertable::into_buffer` validate this. 

I've done the latter in the pull request to pad the `Output` data to 2 bytes: https://github.com/Trangar/artnet_protocol/blob/fix_output_length_and_data/src/command/output.rs#L49-L119

As for the macros, there are 2 major ones:
- Implement `Convertable` on a lot of types (like [these arrays](https://github.com/Trangar/artnet_protocol/blob/master/src/convert.rs#L113-L120)).
- `data_structure`, which adds 2 functions to a struct.
  - [to_bytes](https://github.com/Trangar/artnet_protocol/blob/master/src/macros.rs#L21-L31) sets up a buffer and calls `Convertable` on each field
  - [from](https://github.com/Trangar/artnet_protocol/blob/master/src/macros.rs#L34-L46) calls `Convertable::from_cursor` for each field, and constructs an instance of the struct if all the fields deserialize properly.

**Comment by Firionus** [comment]
> I'm not convinced many people use this crate.

Someone is ;). He has some nice videos on his facebook, too. 
https://www.reddit.com/r/rust/comments/ead99d/gled_animation_software_for_light_installations/

As for the u16 conversion, in my crate I used the `std::convert::From` trait for never-failing conversion (from u8) and `std::convert::TryFrom` for conversions that could fail:
https://github.com/Firionus/artnet_packer/blob/03b124973447d35f6042015f94b6951f419cbb41/src/types/port_address.rs#L19-L54

I was under the impressions these traits would be the most idiomatic way of providing such a feature as they force the end user to handle possible errors of the conversion (compared to panicking in a `new` function, which does not force the user to handle the possible error - his application may just panic). Also, `std` automatically provides `Into` and `TryInto`.

**Comment by VictorKoenders** [comment]
> Someone is

Oh neat people actually use my code

> I was under the impression these traits would be the most idiomatic way

Your impression is right, I always forget these traits exist

**Comment by Firionus** [comment]
Cool, so I'll just port over `PortAddress` from my repo to this one. 

The name comes directly from the Art-Net Spec. It is unfortunate though because it is easily mixed up with UDP/TCP ports. 
We can't use `SubNet` either becasue that is what Art-Net calls Bits 4-7 of the Port Address. We can't use `Universe`, because that is Bits 0-3. 

I'll put in a PR when I'm done (or come back with more questions ;).

**Comment by VictorKoenders** [comment]
Closed in #10 and #8
